chainio/txmgr/txmgr.go

```diff
@@ -0,0 +1,240 @@
+
+type TxManager interface {
+	// Send is used to send a transaction
+	// It takes an unsigned transaction and then signs it before sending
+	// It also takes care of nonce management and gas estimation
+	Send(ctx context.Context, tx *types.Transaction) (*types.Receipt, error)
+
+	// Call is used to call a contract method
+	Call(ctx context.Context, msg ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
+
+	// GetNoSendTxOpts This generates a noSend TransactOpts so that we can use
+	// this to generate the transaction without actually sending it
+	GetNoSendTxOpts() (*bind.TransactOpts, error)
+}
+
+type EthBackend interface {
+	bind.ContractCaller
+	bind.ContractTransactor
+	bind.ContractFilterer
+
+	// TransactionReceipt queries the backend for a receipt associated with
+	// txHash. If lookup does not fail, but the transaction is not found,
+	// nil should be returned for both values.
+	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+
+	ChainID(ctx context.Context) (*big.Int, error)
+}
+
+type SimpleTxManager struct {
+	backend EthBackend
+
+	signerFn signerv2.SignerFn
+
+	log logging.Logger
+
+	from common.Address
+
+	contracts map[common.Address]*bind.BoundContract
+}
+
+var _ TxManager = (*SimpleTxManager)(nil)
+
+func NewSimpleTxManager(
+	backend EthBackend,
+	log logging.Logger,
+	signerFn signerv2.SignerFn,
+	from common.Address,
+) *SimpleTxManager {
+	return &SimpleTxManager{
+		backend:   backend,
+		log:       log,
+		signerFn:  signerFn,
+		from:      from,
+		contracts: map[common.Address]*bind.BoundContract{},
+	}
+}
+
+func (m *SimpleTxManager) Send(ctx context.Context, tx *types.Transaction) (*types.Receipt, error) {
+
+	// Estimate gas and nonce
+	m.log.Debug("Estimating gas and nonce", "tx", tx.Hash().Hex())
+	tx, err := m.estimateGasAndNonce(ctx, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	m.log.Debug("Getting signer for tx", "tx", tx.Hash().Hex())
+	signer, err := m.signerFn(ctx, m.from)
+	if err != nil {
+		return nil, err
+	}
+
+	m.log.Debug("Sending transaction", "tx", tx.Hash().Hex())
+	opts := &bind.TransactOpts{
+		From:      m.from,
+		Nonce:     new(big.Int).SetUint64(tx.Nonce()),
+		Signer:    signer,
+		Value:     tx.Value(),
+		GasFeeCap: tx.GasFeeCap(),
+		GasTipCap: tx.GasTipCap(),
+		GasLimit:  tx.Gas(),
+		Context:   ctx,
+	}
+
+	contract := m.contracts[*tx.To()]
+	// if the contract has not been cached
+	if contract == nil {
+		// create a dummy bound contract tied to the `to` address of the transaction
+		contract = bind.NewBoundContract(*tx.To(), abi.ABI{}, m.backend, m.backend, m.backend)
+		// cache the contract for later use
+		m.contracts[*tx.To()] = contract
+	}
+
+	tx, err = contract.RawTransact(opts, tx.Data())
+	if err != nil {
+		return nil, fmt.Errorf("send: failed to send txn: %w", err)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	receipt := m.waitForTx(ctx, tx)
```

Comment:
unclear to me we want to wait for tx here. What's the reason for this? Why don't we leave that as an option to caller? Or maybe have 2 functions, one that waits for tx to be included and returns a receipt, and another one that just sends it, and returns the tx hash.

Reply:
We want our backend to be robust. Eventually in future we could have retries based on this. I don't want to leave it for callers to retry - their intention is to successfully send tx and not worry about retry or waiting.

Reply:
discussed offline